ios-base/Navigators/Navigator.swift

```diff
@@ -0,0 +1,204 @@
+   */
+  func popTo(route: Route, animated: Bool)
+
+  /**
+   Dismiss your current ViewController.
+   - Parameters:
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func dismiss(animated: Bool, completion: (() -> Void)?)
+}
+
+public extension Navigator {
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool = true, completion: (() -> Void)? = nil) {
+    let viewController = route.screen
+    switch transition {
+    case .modal:
+      currentViewController?.present(viewController, animated: animated, completion: completion)
+      currentViewController = viewController
+    case .push:
+      rootViewController?.pushViewController(viewController, animated: animated)
+      currentViewController = viewController
+    case .reset:
+      rootViewController?.setViewControllers([viewController], animated: animated)
+      currentViewController = viewController
+    case .changeRoot:
+      UIView.animate(withDuration: 0.3) { [weak self] in
+        if let navigation = (viewController as? UINavigationController) {
+          UIApplication.shared.keyWindow?.rootViewController = navigation
+          self?.rootViewController = navigation
+          self?.currentViewController = self?.rootViewController?.topViewController
+        } else {
+          let navigationController = UINavigationController(rootViewController: viewController)
+          UIApplication.shared.keyWindow?.rootViewController = navigationController
+          self?.rootViewController = navigationController
+          self?.rootViewController?.isNavigationBarHidden = true
+          self?.currentViewController = viewController
+        }
+      }
+    }
+  }
+
+  func navigate(to router: Navigator, animated: Bool, completion: (() -> Void)?) {
+    guard let viewController = router.rootViewController else {
+      assert(false, "Navigator does not have a root view controller")
+      return
+    }
+
+    currentViewController?.present(viewController, animated: animated, completion: completion)
+    currentViewController = viewController
+  }
+
+  func pop(animated: Bool = true) {
+    rootViewController?.popViewController(animated: animated)
+    currentViewController = rootViewController?.visibleViewController
+  }
+
+  func popToRoot(animated: Bool = true) {
+    rootViewController?.popToRootViewController(animated: animated)
+  }
+
+  func popTo(index: Int, animated: Bool = true) {
+    guard
+      let viewControllers = rootViewController?.viewControllers,
+      viewControllers.count > index
+      else { return }
+    let viewController = viewControllers[index]
+    rootViewController?.popToViewController(viewController, animated: animated)
+    currentViewController = viewController
+  }
+
+  func popTo(route: Route, animated: Bool = true) {
+    guard
+      let viewControllers = rootViewController?.viewControllers,
+      let viewController = viewControllers.first(where: {
```

Comment:
you can loose the `where` here.

Reply:
The compiler fails 🤔 I think the `first { ... }` works only for equatable (not sure though)

Reply:
That doesn't sound right, it is true for `firstIndex(of:)`. anyways not a big deal